### PR TITLE
test: Add integration test cases for building terraform projects with lambda layers support

### PR DESCRIPTION
**Pull request by moelasmar**
#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [X] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/main.tf [hunk 1]
```diff
@@ -179,4 +185,141 @@ resource "null_resource" "sam_metadata_aws_lambda_layer_version_from_local" {
...
+## */
+
+## /* hello world code builder
+resource "null_resource" "build_hello_world_lambda_function" {
+    triggers = {
+        build_number = "${timestamp()}"
+    }
+
+    provisioner "local-exec" {
+        command = "./py_build.sh \"${local.hello_world_function_src_path}\" \"${local.building_path}\" \"${local.hello_world_artifact_file_name}\" Function"
```
Comment (mildaniel) [inline tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/main.tf:197]: Would be nice to have integration tests that use Lambda Builders in the future too

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/py_build.sh [hunk 2]
```diff
@@ -1,14 +1,22 @@
 src_code=$1
 build_path=$2
 output_name=$3
-echo "building ${src_code} into ${build_path}"
```
Comment (mildaniel) [inline tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/py_build.sh:4]: Do these build scripts work on Windows too?

**Review by mildaniel — APPROVED** [review]

File: samcli/hook_packages/terraform/hooks/prepare/resource_linking.py [hunk 3]
```diff
@@ -207,8 +207,15 @@ def _build_module_resources_from_configuration(module_configuration: Dict, modul
 
         expressions = config_resource.get("expressions", {})
         for expression_name, expression_value in expressions.items():
+            # we do not process the attributes of type dictionary
+            # Todo add dictionary type attributes post beta
+            if isinstance(expression_value, list):
+                LOG.debug("Skip processing the attribute %s as its value is a map.", expression_name)
```
Comment (sriram-mv) [inline samcli/hook_packages/terraform/hooks/prepare/resource_linking.py:213]: nit: we check for list and have debug message on "map"?

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/artifacts/HelloWorldFunction/app.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import json
+import lib
+import boto3
+
+# import requests
```
Comment (sriram-mv) [inline tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/artifacts/HelloWorldFunction/app.py:5]: something that snuck in?

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_s3_backend/artifacts/layer1/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+boto3
+ptvsd
```
Comment (sriram-mv) [inline tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_s3_backend/artifacts/layer1/requirements.txt:2]: why ptvsd? its for debugging right?

**Review by sriram-mv — APPROVED** [review]
had some questions

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/py_build.sh (hunk 2 above)
Comment (moelasmar) [reply tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/py_build.sh:4]: I already discussed this point with @hawflau, and he will add a new project for the windows.

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_s3_backend/artifacts/layer1/requirements.txt (hunk 5 above)
Comment (moelasmar) [reply tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_s3_backend/artifacts/layer1/requirements.txt:2]: I just add some dependencies, and then imported in the python code to make sure that the build is really working.

File: samcli/hook_packages/terraform/hooks/prepare/resource_linking.py (hunk 3 above)
Comment (moelasmar) [reply samcli/hook_packages/terraform/hooks/prepare/resource_linking.py:213]: This is how it works in Terraform, the customer defines an attribute and can set its value to a map, but in the Terraform state,  it represented as list of dictionaries.

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_s3_backend/artifacts/layer1/requirements.txt (hunk 5 above)
Comment (moelasmar) [reply tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_s3_backend/artifacts/layer1/requirements.txt:2]: and actually it seems no used in the library code, I will remove it.

File: tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/artifacts/HelloWorldFunction/app.py (hunk 4 above)
Comment (moelasmar) [reply tests/integration/testdata/buildcmd/terraform/zip_based_lambda_functions_local_backend/artifacts/HelloWorldFunction/app.py:5]: I will clean it

**Pull request merged**